### PR TITLE
Add a new type of switching function with an adjustable weight parameter

### LINKED ISSUE #9042: Add an new option in SwitchingFunctionMaterial
### Description of the enhancement or error report
The new form of switch function h will be a general formula of a 234-order polynomial, with a new adjustable parameter. The new function should also satisfy the following constraints:
h(0)=0, h(1)=1
h'(0)=h'(1)=0

### Rationale for the enhancement or information for reproducing the error
Current options (SIMPLE/HIGH) only include 23-order and 345-order polynomial, which is not capable to describe the symmetry breaking during the phase transitions in crystalline solids in multiphase model. To make a general mathematical formula capturing the physical nature of the phase transition and being consistent with Landau's classical theory, a flexible 234-order polynomial is required. 

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
It will affect related multiphase model component when the new option is applied.

### PR DESCRIPTION
**Pull request by yipenggao**
Add a new type of switching function (MIX) with an adjustable weight parameter (mix_weight).

Add a new inputfile (modified from DerivativeMultiPhaseMaterial.i) to test the new function.

closes #9042

### CODE REVIEW

**Comment by yipenggao** [comment]
This my first time to change the source code file. Please let me know if it is not within the code standard. :)

Thanks.

**Comment by yipenggao** [comment]
Ok. It fails all the checks.... I will work on it next weekend... :(

**Comment by dschwen** [comment]
You need to activate the new test by putting it into the tests file. I have more comments later.

**Comment by yipenggao** [comment]
@dschwen
Should I put create a new folder under /tests/ and put the input and output file there? Also a file named "tests" is needed, right?

**Comment by dschwen** [comment]
No new directory. Look where the other modes of the switching function are tested and make your test input similar to those.

File: modules/phase_field/src/materials/SwitchingFunctionMaterial.C [hunk 1]
```diff
@@ -11,18 +11,24 @@ InputParameters
...
-                             "eta^3*(6*eta^2-15*eta+10)");
-  MooseEnum h_order("SIMPLE=0 HIGH", "SIMPLE");
+                             "eta^3*(6*eta^2-15*eta+10)\nMIX: "
+                             "mix_weight*(3*eta^2-2*eta^3)+(1-mix_weight)(4*eta^3)-3*eta^4");
+  MooseEnum h_order("SIMPLE=0 HIGH MIX", "SIMPLE");
   params.addParam<MooseEnum>(
       "h_order", h_order, "Polynomial order of the switching function h(eta)");
   params.set<std::string>("function_name") = std::string("h");
+
+  Real mix_weight = 1;
```
Comment (dschwen) [inline modules/phase_field/src/materials/SwitchingFunctionMaterial.C]: remove this variable

File: modules/phase_field/src/materials/SwitchingFunctionMaterial.C [hunk 2]
```diff
@@ -46,6 +52,14 @@ SwitchingFunctionMaterial::computeQpProperties()
       _prop_d2f[_qp] = n * (120.0 * n * n - 180.0 * n + 60.0);
       break;
 
+    case 2: // MIX
+      _mix_weight = _mix_weight > 1 ? 1 : (_mix_weight < 0 ? 0 : _mix_weight);
```
Comment (dschwen) [inline modules/phase_field/src/materials/SwitchingFunctionMaterial.C]: No. Range limit the parameter, set it in the initializer list, and naver change it again. In fact make it `const` in the declaration.

**Comment by dschwen** [comment]
No new directory. Look where the other modes of the switching function are tested and make your test input similar to those.

**Comment by dschwen** [comment]
I'd actually rather see this implemented in a derived class. The mix weight parameter is only used for the one new h order option.

**Comment by dschwen** [comment]
Why did you close this?

**Comment by yipenggao** [comment]
@dschwen 
... Sorry, I reset it to previous version (then push....), and right now I am working on adding a new class files. I didn't realize that the PR closed automatically...

Can we reopen it or we have to make a new PR?

**Comment by yipenggao** [comment]
It still fails the check... Any hint about why?

**Comment by permcody** [comment]
Click on the precheck target details. It'll take you to CIVET where you can open the results to look at them. You just need a ticket reference in one or more of your commits in this PR.

**Comment by permcody** [comment]
Also we prefer "rebase" strategy with MOOSE development, not "merge" strategy.

**Comment by yipenggao** [comment]
@permcody 
Thanks for your help!

Most of the fail information are "Missing gold file".

I have put the inputfile in .../phase_field/tests/MultiPhase/ , and the output file in .../phase_field/tests/MultiPhase/gold/

Is that the correct way to do it?

File: modules/phase_field/tests/MultiPhase/tests [hunk 3]
```diff
@@ -65,4 +65,9 @@
     input = 'switchingfunctionmultiphasematerial.i'
     exodiff = 'switchingfunctionmultiphasematerial_out.e'
   [../]
+  [./DerivativeMultiPhaseMaterial_SwitchingMix]
+    type = 'Exodiff'
+    input = 'DerivativeMultiPhaseMaterial_SwitchingMix.i'
+    exodiff = 'DerivativeMultiPhaseMaterial_SwithcingMix_out.e'
```
Comment (permcody) [inline modules/phase_field/tests/MultiPhase/tests]: The problem is spelling 'Swithcing' -> 'Switching'

**Comment by yipenggao** [comment]
@permcody 

Many thanks for your review. I will be more careful.

**Comment by yipenggao** [comment]
It is said "Failed(exodiff)".

Is that because I use incorrect type option in the tests file?
type = 'Exodiff'

**Comment by permcody** [comment]
No, it's because it actually is diffing. Scroll up in the output and look at the differences. They are very tiny (in the e-10 range) which is right near our absolute zero value. The Exodiff tolerances can be adjusted but you generally shouldn't change them. The defaults are listed in the Exodiff object itself. There's a relative tolerance and an absolute floor.

Instead you might re-evaluate your test and decide whether or not you need that many steps to actually test your physics (over 30). The first 20 or so are identical (this is all in the output). You start to get a little numerical round-off as the small differences accumulate. There are many ways to make all of these things go away. If you have any questions about the Exodiff output don't hesitate to ask. It's not too difficult to look through though. 

What you are experiencing is common by the way. You are probably creating tests on a Mac but the test systems are mostly Linux.

**Comment by permcody** [comment]
@yipenggao - As you shorten up the test, you'll need to update the gold file too. Exodiff complains if the number of output steps don't match.

File: modules/phase_field/include/materials/SwitchingFunctionMaterialMix.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+InputParameters validParams<SwitchingFunctionMaterialMix>();
+
+/**
+ * Material class to provide the switching function \f$ h(\eta) \f$ for
+ * the KKS system.
+ *
+ * \see KKSPhaseChemicalPotential
+ * \see KKSCHBulk
+ */
+class SwitchingFunctionMaterialMix : public OrderParameterFunctionMaterial
```
Comment (dschwen) [inline modules/phase_field/include/materials/SwitchingFunctionMaterialMix.h]: Please rename to `MixedSwitchingFunctionMaterial`; this should end on Material for consistency.

File: modules/phase_field/include/materials/SwitchingFunctionMaterialMix.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  SwitchingFunctionMaterialMix(const InputParameters & parameters);
+
+protected:
+  virtual void computeQpProperties();
+
+  /// Polynomial order of the switching function \f$ h(\eta) \f$
+  MooseEnum _h_order;
+
+  /// Weight parameter of mixed-type h(eta)
+  Real _weight;
```
Comment (dschwen) [inline modules/phase_field/include/materials/SwitchingFunctionMaterialMix.h]: make `const`

File: modules/phase_field/src/materials/SwitchingFunctionMaterialMix.C [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+template <>
+InputParameters
+validParams<SwitchingFunctionMaterialMix>()
+{
+  InputParameters params = validParams<OrderParameterFunctionMaterial>();
+  params.addClassDescription("Helper material to provide h(eta) and its derivative in one of two "
+                             "polynomial forms.\nMIX234: "
+                             "weight*(3*eta^2-2*eta^3)+(1-weight)(4*eta^3)-3*eta^4");
+  MooseEnum h_order("MIX234=0", "MIX234");
+  params.addParam<MooseEnum>(
```
Comment (dschwen) [inline modules/phase_field/src/materials/SwitchingFunctionMaterialMix.C]: remove the `h_order` parameter. No point in having a single option parameter.

File: modules/phase_field/src/materials/SwitchingFunctionMaterialMix.C [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+{
+}
+
+void
+SwitchingFunctionMaterialMix::computeQpProperties()
+{
+  Real n = _eta[_qp];
+  n = n > 1 ? 1 : (n < 0 ? 0 : n);
+
+  switch (_h_order)
```
Comment (dschwen) [inline modules/phase_field/src/materials/SwitchingFunctionMaterialMix.C]: if `_h_order` is removed the switch block becomes unnecessary

File: modules/phase_field/tests/MultiPhase/DerivativeMultiPhaseMaterial_SwitchingMix.i [hunk 8]
```diff
@@ -0,0 +1,385 @@
+[Mesh]
```
Comment (dschwen) [inline modules/phase_field/tests/MultiPhase/DerivativeMultiPhaseMaterial_SwitchingMix.i]: model you test based on `switchingfunction3phasematerial.i` for example. You do not need a solve to test this material. Please try to keep to the standard for filename capitalization and use of underscores for at least the directory the file is located in.

**Review by dschwen — CHANGES_REQUESTED** [review]

**Comment by dschwen** [comment]
Oh, right, before this becomes mergeable it needs a documentation file! I can show you how to get going on that.

**Comment by dschwen** [comment]
When you apply the style fixes please _squash_ your commits as much as it makes sense.

**Comment by yipenggao** [comment]
Great! It finally passes all checks.

Then how to make documentation file. Any online instruction?

File: docs/content/documentation/systems/Materials/phase_field/MixedSwitchingFunctionMaterial.md [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+!description /Materials/MixedSwitchingFunctionMaterial
+
+$$
+k(3\eta^2-2\eta^3)+(1-k)(4\eta^3-3\eta^4)
+$$
+
+$$
+k(2\eta^2-\eta^4)+(1-k)(3\eta^4-2\eta^6)
+$$
+
```
Comment (dschwen) [inline docs/content/documentation/systems/Materials/phase_field/MixedSwitchingFunctionMaterial.md]: please at least mention MIX234 and MIX246 here. Just two terms is a bit weak for documentation.

**Review by dschwen — APPROVED** [review]

**Pull request merged**